Integrations/Visual Studio (#2297)
conan-io/docs: integrations/visual_studio/creating_packages.rst

.. code-block:: bash

    $ Hello World Debug!
    # Switch IDE to Release mode, repeat
    $ Hello World Release!

Because the ``hello.cpp`` file contains an ``#ifdef _DEBUG`` to switch between debug and release
message.

In the repository, there is already a ``conanfile.py`` recipe:

.. code-block:: python

    from conans import ConanFile
    from conan.tools.microsoft import MSBuild

    class HelloConan(ConanFile):
        name = "Hello"
        version = "0.1"
        license = "MIT"
        url = "https://github.com/conan-io/examples"
        settings = "os", "compiler", "build_type", "arch"
        exports_sources = "src/*", "build/*.vcxproj*", "build/*.sln*"

        def build(self):
            msbuild = MSBuild(self)
            msbuild.build("build/HelloLib/HelloLib.sln")

        def package(self):
            self.copy("*.h", dst="include", src="src")
            self.copy("*.lib", dst="lib", keep_path=False)
            self.copy("*.dll", dst="bin", keep_path=False)

        def package_info(self):
            self.cpp_info.libs = ["HelloLib"]

This recipe is using the (:ref:)`MSBuild() build helper <msbuild>` to build the ``sln`` project.

> If our recipe has ``requires``, the ``MSBUILD`` helper will also take care of inject all the needed

Reviewer: This is wrong, the ``MSBuild`` helper doesn't inject all of that, it is mostly dummy, a thin wrapper around the command line. That responsibility is not part of ``MSBuildDeps`` and ``MSBuildToolchain`` (better leave the explanation when you use them)

Author: Done

to allow our project to locate the declared dependencies.

The recipe contains also a ``test_package`` folder with a simple example consuming application.
In this example, the consuming application is using CMake to build, but it could also use Visual
Studio too. We have left the CMake one because it is the default generated with
:command:`conan new`, and also to show that packages created from Visual Studio projects can also
be consumed with other build systems like CMake.

Once we want to create a package, it is advised to close VS IDE, clean the temporary build files
from VS to avoid problems, then create and test the package. Here it is using system defaults,
assuming they are Visual Studio 19, Release, x86_64:

.. code-block:: bash

   # close VS
   $ git clean -xdf
   $ conan create . demo/testing
   ...
   > Hello World Release!

Instead of closing the IDE and running the command:`git clean` we could also configure a smarter
filter in ``exports_sources`` field, so temporary build files are not exported into the recipe.

This process can be repeated to create and test packages for different configurations:

.. code-block:: bash

   $ conan create . demo/testing -s arch=x86

>    $ conan create . demo/testing -s compiler="Visual Studio" -s compiler.runtime=MDd -s build_type=Debug

Reviewer: Lets move all examples to the new ``msvc`` compiler setting, it is extremely likely that ``Visual Studio`` will be removed in 2.0

Author: Done

dependencies.

Now you can open the IDE and build and run the app (by the way, the chat function is just calling
the ``hello()`` function two or three times, depending on the build type):

.. code-block:: bash

    $  build\ChatLib\ChatLib.sln
    # Switch to Release
    # MyChat -> Set as Startup Project
    # Ctrl + F5 (Debug -> Run without debugging)
    > Hello World Release!
    > Hello World Release!

If you wish to link with the debug version of Hello package, just install it and change IDE build
type:

.. code-block:: bash

    $ conan install . -s build_type=Debug -s compiler="Visual Studio" -s compiler.runtime=MDd
    # Switch to Debug
    # Ctrl + F5 (Debug -> Run without debugging)
    > Hello World Debug!
    > Hello World Debug!
    > Hello World Debug!

Now you can close the IDE and clean the temporary files:

.. code-block:: bash

    # close VS IDE
    $ git clean -xdf

Again, there is a ``conanfile.py`` package recipe in the repository, together with a
``test_package``. The recipe is almost identical to the above one, just with two minor
differences:

.. code-block:: python

    requires = "hello/0.1@demo/testing"
    ...

>     generators = "visual_studio"

Reviewer: Wrong, legacy generator. I strongly recommend actually running this docs and the example code against the Conan 2.0-alpha

Author: The examples in https://github.com/conan-io/examples are updated and tested. I just forgot to update this snippet. Fixed.

.. _visual_studio_packages:

> |visual_logo| Creating and reusing packages based on Visual Studio

Reviewer: ```suggestion
|visual_logo| Creating and reusing packages with Visual Studio
```

.. _visual_studio_packages:

|visual_logo| Creating and reusing packages based on Visual Studio
==================================================================

Conan has different helpers to manage Visual Studio and MSBuild based projects.
This how-to illustrates how to put them together to create and consume packages that are purely
based on Visual Studio. This how-to is using VS2015, but other versions can be used too.

Creating packages
------------------

Start cloning the existing example repository, containing a simple "Hello World" library, and
application:

.. code-block:: bash

>     $ git clone https://github.com/memsharded/hello_vs

Reviewer: There is work to do here, because this is in my personal repo. Must be moved to "examples" repo first

    $ cd hello_vs

It contains a ``src`` folder with the source code and a ``build`` folder with a Visual Studio 2015
solution, containing 2 projects: the HelloLib static library, and the Greet application. Open it:

.. code-block:: bash

    $ build\HelloLib\HelloLib.sln

You should be able to select the ``Greet`` subproject -> ``Set as Startup Project``.
Then build and run the app with Ctrl+F5. (Debug -> Start Without Debugging)

.. code-block:: bash

    $ Hello World Debug!
    # Switch IDE to Release mode, repeat
    $ Hello World Release!

Because the ``hello.cpp`` file contains an ``#ifdef _DEBUG`` to switch between debug and release
message.

In the repository, there is already a ``conanfile.py`` recipe:

.. code-block:: python

>     from conans import ConanFile, MSBuild

Reviewer: Legacy MSBuild integration, not the modern one

assuming they are Visual Studio 14, Release, x86_64:

.. code-block:: bash

   # close VS
   $ git clean -xdf
   $ conan create . memsharded/testing
   ...
   > Hello World Release!

Instead of closing the IDE and running the command:`git clean` we could also configure a smarter
filter in ``exports_sources`` field, so temporary build files are not exported into the recipe.

This process can be repeated to create and test packages for different configurations:

.. code-block:: bash

   $ conan create . memsharded/testing -s arch=x86
   $ conan create . memsharded/testing -s compiler="Visual Studio" -s compiler.runtime=MDd -s build_type=Debug
   $ conan create . memsharded/testing -s compiler="Visual Studio" -s compiler.runtime=MDd -s build_type=Debug -s arch=x86

.. note::

    It is not mandatory to specify the ``compiler.runtime`` setting. If it is not explicitly
    defined, Conan will automatically use ``runtime=MDd`` for ``build_type==Debug`` and
    ``runtime=MD`` for ``build_type==Release``.

You can list the different created binary packages:

.. code-block:: bash

    $ conan search hello/0.1@memsharded/testing

> Uploading binaries

Reviewer: This section can be dropped, nothing really useful or specific here

Author: Dropped

------------------

Your locally created packages can already be uploaded to a Conan remote.
If you created them with the original username "memsharded", as from the git clone, you might want
to do a :command:`conan copy` to put them on your own username. Of course, you can also directly
use your user name in :command:`conan create`.

Another alternative is to configure the permissions in the remote, to allow uploading packages
with different usernames. By default, Artifactory will do it but Conan server won't: Permissions
must be given in the ``[write_permissions]`` section of *server.conf* file.

Reusing packages
----------------

> To use existing packages directly from Visual Studio, Conan provides the ``visual_studio``

Reviewer: Legacy generator too
